### PR TITLE
SAML: Add client_id to service provider

### PR DESCRIPTION
**Pull request by tung2744**
ref DEV-2002

Steps to update existing configs:
1. Deploy the code in this pr
2. Add `client_id` in existing config. Old endpoints should still work, but metadata will be updated.
3. Switch to new endpoints.
4. Remove `id` from configs.
5. Remove `id` from code.

### CODE REVIEW

File: pkg/lib/config/config.go [hunk 1]
```diff
@@ -357,6 +357,22 @@ func (c *AppConfig) validateSAML(ctx *validation.Context) {
...
+						found = true
+						break
+					}
+				}
+				if !found {
+					ctx.Child("saml", "service_providers", strconv.Itoa(idx), "client_id").
+						EmitErrorMessage("client_id does not exist in /oauth/clients")
+				}
+			}
+		}
```
Comment (louischan-oursky) [inline pkg/lib/config/config.go:375]: Add tests for this validation logic.

File: pkg/lib/config/saml.go [hunk 2]
```diff
@@ -37,7 +37,15 @@ var _ = Schema.Add("SAMLServiceProviderConfig", `
...
-	"required": ["acs_urls"]
+	"required": ["acs_urls"],
+	"anyOf": [
+		{
+			"required": ["id"]
+		},
+		{
+			"required": ["client_id"]
+		}
+	]
```
Comment (louischan-oursky) [inline pkg/lib/config/saml.go:48]: Add tests for this validation.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

**Comment by tung2744** [comment]
Added, thanks!

File: pkg/lib/config/testdata/config_tests.yaml [hunk 3]
```diff
@@ -1200,3 +1200,39 @@ config:
...
+  id: test
+  http:
+    public_origin: http://test
+  saml:
+    signing:
+      key_id: asdfasdf
+    service_providers:
+      - nameid_format: urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress
+        acs_urls:
+          - http://localhost:3000/acs
```
Comment (louischan-oursky) [inline pkg/lib/config/testdata/config_tests.yaml:1221]: Add two more clients here to show the validation pass case. That is
1. A SP with `id` should pass
2. A SP with `client_id` should pass

File: pkg/lib/config/testdata/config_tests.yaml [hunk 4]
```diff
@@ -1200,3 +1200,39 @@ config:
...
+  http:
+    public_origin: http://test
+  saml:
+    signing:
+      key_id: asdfasdf
+    service_providers:
+      - client_id: notexist
+        nameid_format: urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress
+        acs_urls:
+          - http://localhost:3000/acs
```
Comment (louischan-oursky) [inline pkg/lib/config/testdata/config_tests.yaml:1238]: Add one more client here to show the validation pass case. That is
1. A SP with a client id pointing to an existing client.

Then one test case is enough to demo the validation is correct.

**Review by louischan-oursky — CHANGES_REQUESTED** [review]

File: pkg/lib/config/testdata/config_tests.yaml (hunk 3 above)
Comment (tung2744) [reply pkg/lib/config/testdata/config_tests.yaml:1229]: Added two items

File: pkg/lib/config/testdata/config_tests.yaml (hunk 4 above)
Comment (tung2744) [reply pkg/lib/config/testdata/config_tests.yaml:1254]: > Then one test case is enough to demo the validation is correct.

Not sure what you mean here. I added a valid item in the above test case which point to a valid client.

Comment (tung2744) [reply pkg/lib/config/testdata/config_tests.yaml:1254]: But I think we still need another test case to show there is an error if `client_id` is not a valid client.

Comment (louischan-oursky) [reply pkg/lib/config/testdata/config_tests.yaml:1254]: Your test`saml-service-provider-invalid-client-id` did this already.

**Review by louischan-oursky — APPROVED** [review]

**Pull request merged**